### PR TITLE
Mark AppliedType cachedSuper valid Nowhere when using provisional args

### PR DESCRIPTION
**Pull request by EugeneFlesselle**
I'm not sure why it wasn't the case previously. 

It solves some stale caching issues I encounter when introducing more caching for match types in #20268, as well as the problem in `neg/typeclass-encoding3` apparently.

There seems to be no significant performance impact from the reduced opportunities for caching.

### CODE REVIEW

**Comment by EugeneFlesselle** [comment]
Test performance please

**Comment by EugeneFlesselle** [comment]
test performance please

**Comment by EugeneFlesselle** [comment]
test performance please: 59b0f3ae c6fbe6fe

**Comment by EugeneFlesselle** [comment]
The last commit only changes a test case, the previous commit (with the change in AppliedType) looks to be inline with `main`.
Screenshot 2024-06-06 at 17 30 48

**Review by odersky — COMMENTED** [review]
> I'm not sure why it wasn't the case previously.

I think the thinking was because it's only type constructor that influences the shape of the supertype, if the type constructor is no longer provisional, the supertype is always the same. It might still be provisional if one of the arguments is provisional, but that's another question. I am not really sure where the hole in that reasoning is. The change seems to fix something, but i am not sure why. It would be good to have an explanation, why the problem happened and how changing isProvisional fixed it.

**Comment by EugeneFlesselle** [comment]
After some investigation, here is what I found: `TypeApplications#appliedTo` _can_ perform reductions that depend on the arguments, which are not structurally insignificant (although it is true the results are equal with respect to `=:=`).
In particular, it will perform eta-reduction if the de-aliased tycon is an HKTypeLambda and the `dealiased.paramRefs == dealiasedArgs`, but ignoring variance annotations.
This _can_ have an impact on `TypeComparer#compareTypeLambda` as it requires the variances to conform.

In the example of `pos/typeclass-encoding3b`
```scala
object Monad extends TypeClassCompanion1:
  type Impl[T[_]] = MonadCommon { type This = [X] =>> T[X] }

implicit object ListMonad extends MonadCommon:
  type This[+X] = List[X]

def flattened[T[_], A]($this: T[T[A]])(implicit ev: Monad.Impl[T]): T[A] = ???
flattened(List(List(1, 2, 3), List(4, 5)))(using ListMonad)
```
the key observation is that the `This` type member of `ListMonad` has a variance annotation.

When checking `ListMonad <:< functors.Monad.Impl[T]` we eventually get to the comparison `[X] =>> T[X] <:< [+X] =>> List[X]` which fails the variance conformance checks, despite the fact that `T` has been instantiated to `List`, since it has been substituted into the refinement (and cached) before its instantiation.
If we however recompute the `superType` after having found `T := List`, `appliedTo` now yields the same covariance annotation, which allows the type comparison to succeed.

**Comment by EugeneFlesselle** [comment]
If the changes in this PR are deemed sufficient to address the issue, I can definitely add a comment to `AppliedType#superType` explaining why the change is needed.

**Comment by odersky** [comment]
That's a great answer. I think we should have a condensed version in the code for AppliedType#superType, with a link to the test that would otherwise fail.

**Comment by EugeneFlesselle** [comment]
> That's a great answer. I think we should have a condensed version in the code for AppliedType#superType, with a link to the test that would otherwise fail.

Sounds good, I'll do that 👍

**Comment by EugeneFlesselle** [comment]
Note I also added an unrelated change to the doc in https://github.com/scala/scala3/pull/20527/commits/ee5481408b09e95ce9becaea510be8687ff27c87

File: tests/pos/typeclass-encoding3b.scala [hunk 1]
```diff
@@ -345,5 +345,12 @@ object functors {
   }
 
   MonadFlatten.flattened(List(List(1, 2, 3), List(4, 5))) // ok, synthesizes (using ListMonad)
-  MonadFlatten.flattened(List(List(1, 2, 3), List(4, 5)))(using ListMonad) // error
+  MonadFlatten.flattened(List(List(1, 2, 3), List(4, 5)))(using ListMonad) // was an error
+  /*
+  When checking `ListMonad <:< functors.Monad.Impl[T]`
+  we eventually get to the comparison `[X] =>> T[X] <:< [+X] =>> List[X]`
+  because the `This` type member of `ListMonad` has a covariance annotation.
+  This fails the variance conformance checks despite the fact that T has been instantiated to List,
```
Comment (dwijnand) [inline tests/pos/typeclass-encoding3b.scala:353]: I don't follow the comment.  How does it failing the conformance check lead it to not erroring?

**Review by dwijnand — APPROVED** [review]
LGTM, just a question

File: tests/pos/typeclass-encoding3b.scala (hunk 1 above)
Comment (EugeneFlesselle) [reply tests/pos/typeclass-encoding3b.scala:353]: Ah, I'd overlooked updating the tense

**Pull request merged**